[Concept Entry] Postgresql subscripts (#5656)
Codecademy/docs: content/postgresql/concepts/subscripts/subscripts.md

---
Title: 'Subscripts'

> Description: 'In PostgreSQL, subscripts are used to perform various operations on arrays. They offer a concise way to handle and manage array data.'

Reviewer: The definition can be clearer by mentioning the operations that can be performed. Also, the position is the most important aspect of subscripts; we should use it to explain the role of subscripts.

Description can be something like this -

```suggestion
Description: 'Subscripts are used to access, update, or manipulate specific elements of an array by their position or index.'
```

Subjects:
  - 'Computer Science'
  - 'Data Science'
  - 'Web Development'
Tags:
  - 'Arrays'
  - 'Database'
  - 'Data Structures'
  - 'PostgreSQL'
  - 'SQL'
CatalogContent:
  - 'paths/analyze-data-with-sql'
  - 'paths/data-science'
---

**Subscripts** are a fundamental concept in PostgreSQL, used to access and manipulate elements within an array. By using subscripts, it becomes easy to retrieve, update, or remove elements from an array by referring to their specific positions.

## Syntax

For a one-dimensional array, the following syntax is used:

```pseudo
array_name[subscript]
```
- `array_name`: The name of the array being accessed.
- `subscript`: The index position of the element to be accessed within the array.

Alternatively, for multi-dimensional array the following syntax is used:

> ```pseudo
> array_name[subscript_1, subscript_2, ..., subscript_n]

Reviewer: In PostgreSQL, multi-dimensional array elements are accessed using nested square brackets, where each set of brackets corresponds to a specific dimension. Refer [docs](https://www.postgresql.org/docs/current/arrays.html) for more info.

The syntax for accessing multi-dimensional arrays in Postgres is wrong, and the correct syntax is as follows -

```suggestion
```pseudo
array_name[subscript_1][subscript_2]...[subscript_n]
```

```
- `array_name`: The name of the array being accessed.
- `subscript_1, subscript_2, ..., subscript_n`: The index positions for each dimension of the array to access the specific element.

## Example

> Suppose, we have a table `students` with columnname as `grades` that stores a multi-dimensional array of test scores:

Reviewer: This sentence can be a bit more clear as follows -

```suggestion
Suppose we have a table `students` with column name `grades` that stores a multi-dimensional array of test scores:
```

```sql
CREATE TABLE students (
    id SERIAL PRIMARY KEY,
    name VARCHAR(100),
    grades INTEGER[][]
);
```
Insert some data into the table:

```sql
INSERT INTO students (name, grades)
VALUES 
    ('Bryan', ARRAY[[98, 60], [75, 52]]);
```
If we want to access Bryan's score, we will run the below query:

```sql

> SELECT grades[2, 2] AS second_subject_score
> FROM students
> WHERE name = 'Bryan';

Reviewer: In PostgreSQL, for multi-dimensional arrays, multiple subscripts should be used with separate square brackets for each dimension. The correct query should be:

```suggestion
SELECT grades[2][2] AS second_subject_score
FROM students
WHERE name = 'Bryan';
```
